### PR TITLE
Speed up pre-commit hooks with lint-staged

### PR DESCRIPTION
**Pull request by chaance**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code.
- [ ] Add or edit tests to reflect the change (run `yarn test`).
- [ ] Add or edit Storybook examples to reflect the change (run `yarn dev`).
- [ ] Add documentation to support any new features.

This pull request:

- [ ] Fixes a bug in an existing package
- [ ] Adds additional features/functionality to an existing package
- [ ] Updates documentation or example code
- [x] Other

### CODE REVIEW

**Comment by benoitgrelard** [comment]
@chaance So the main speed-up is from the fact that we're only going to re-lint / typecheck the files that are being staged rather than the whole codebase?

**Review by benoitgrelard — COMMENTED** [review]
Looks good!

One small nit: Could we move this into the `"lint-staged"` key in the `package.json` to minimise the proliferation of config files at the root?

File: lint-staged.config.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+const lint = 'eslint --quiet --fix --cache';
+
+// https://github.com/okonet/lint-staged/issues/825#issuecomment-620018284
+module.exports = {
+  '*.{js}': [lint],
+  '*.{ts,tsx}': [lint, () => 'tsc --skipLibCheck --noEmit'],
+};
```
Comment (jjenzz) [inline lint-staged.config.js:7]: I believe we're able to put this on `package.JSON` right? Just thinking if configs are small perhaps they're fine there? We have others there at least.

**Review by jjenzz — CHANGES_REQUESTED** [review]
Love this. Been meaning to chat about this. Thoughts on moving it to push rather than commit too?

**Comment by chaance** [comment]
@benoitgrelard See the comment I left in the config file; in order for `tsc` to work properly with `lint-staged` (and for our scripts to run cross platform) one of those keys we have to use the function syntax in a js config file. I kind of hate it too but I couldn't get it to work in package.json without running bash in the command, which wouldn't work on Windows and would exclude contributors. 😕

**Comment by chaance** [comment]
> Love this. Been meaning to chat about this. Thoughts on moving it to push rather than commit too? 

Even better IMO!

File: lint-staged.config.js (hunk 1 above)
Comment (benoitgrelard) [reply lint-staged.config.js:7]: haha I left exactly the same comment, but it seems there's an issue with that, hence why the comment above (I missed it too)

**Comment by benoitgrelard** [comment]
> > Love this. Been meaning to chat about this. Thoughts on moving it to push rather than commit too?
> 
> Even better IMO!

We can't really move this to pre-push hook right? otherwise how would the `--fix` work?

**Comment by chaance** [comment]
@benoitgrelard I'm thinking we could keep `prettier` and `eslint` in pre-commit. Those actually run pretty fast anyway when only executed on staged changes. `tsc` is significantly slower and I think we can move that to pre-push.

**Comment by chaance** [comment]
@benoitgrelard @jjenzz Updated a bit. Per my comment above I kept `prretty-quick` and `eslint` in the pre-commit hook, runs only on staged files and is way faster now. `tsc` goes in pre-push because it was much slower, and that let me delete the config file and keep everything in `package.json`.

File: package.json [hunk 2]
```diff
@@ -9,6 +9,7 @@
   ],
   "scripts": {
     "lint": "eslint packages --ext .js,.ts,.tsx",
+    "types:check": "tsc --skipLibCheck --noEmit",
```
Comment (chaance) [inline package.json:12]: Putting the script here instead of running `tsc` directly from husky is what let me delete the config file with function syntax.

**Review by benoitgrelard — APPROVED** [review]
Even better!

**Comment by jjenzz** [comment]
🎉 :+1:

**Pull request merged**